uro98/assignment3/FindAllWords.java

```diff
@@ -0,0 +1,80 @@
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class FindAllWords {
+  
+  /**
+   * Find all the words in the dictionary that can be formed in the grid.
+   * @param grid
+   * @param dictionary
+   * @return
+   */
+  public static Set<String> findAllWords(Grid grid, Dictionary dictionary) {
+    Set<String> totalWords = new HashSet<String>();
+    
+    // For each position in the grid, find all the words that can be formed starting from there
+    for (int i = 0; i < grid.getHeight(); i++) {
+      for (int j = 0; j < grid.getWidth(); j++) {
+        Set<String> wordSet = FindAllWords.searchWords(
+            grid,
+            dictionary,
+            new boolean[grid.getHeight()][grid.getWidth()],
+            new HashSet<String>(),
+            new ArrayList<String>(),
+            i,
+            j);
+        totalWords.addAll(wordSet);
+      }
+    }
+    
+    return totalWords;
+  }
+  
+  public static Set<String> searchWords(
+      Grid grid,
+      Dictionary dictionary,
+      boolean[][] visited,
+      Set<String> words,
+      List<String> path,
```

Comment:
It is not necessary to store `String`s as path elements. `Character` would be enough.

Even better - just use `String` rather than `List<String>` - this will simplify code quite a bit since you will not need to join elements and make explicit copies.